pallets/gear/src/manager/journal.rs

```diff
@@ -272,21 +286,37 @@ where
         message_id: MessageId,
         program_id: ProgramId,
         awakening_id: MessageId,
+        delay: u32,
     ) {
-        if let Some(dispatch) = Pallet::<T>::wake_dispatch(
-            program_id,
-            awakening_id,
-            MessageWokenRuntimeReason::WakeCalled.into_reason(),
-        ) {
-            QueueOf::<T>::queue(dispatch)
-                .unwrap_or_else(|e| unreachable!("Message queue corrupted! {:?}", e));
-        } else {
-            log::debug!(
-                "Attempt to wake unknown message {:?} from {:?}",
+        if delay.is_zero() {
+            if let Some(dispatch) = Pallet::<T>::wake_dispatch(
+                program_id,
                 awakening_id,
-                message_id
-            );
+                MessageWokenRuntimeReason::WakeCalled.into_reason(),
+            ) {
+                QueueOf::<T>::queue(dispatch)
+                    .unwrap_or_else(|e| unreachable!("Message queue corrupted! {:?}", e));
+
+                return;
+            }
+        } else if WaitlistOf::<T>::contains(&program_id, &awakening_id) {
+            let expected_bn =
+                SystemPallet::<T>::block_number().saturating_add(delay.unique_saturated_into());
+            let task = ScheduledTask::WakeMessage(program_id, awakening_id);
+
+            if !TaskPoolOf::<T>::contains(&expected_bn, &task) {
```

Comment:
please state in the comment that the check is needed for the cross-block protection against double wake